src/lib.rs

```diff
@@ -160,22 +183,61 @@ impl Drop for IRRootSignature {
     }
 }
 
+#[derive(Error, Debug)]
+pub enum RootSignatureError {
+    #[error("Failed to create IRRootSignature: {0:?}")]
+    CreateError(ffi::IRErrorCode),
+    #[error("Failed to create IRRootSignature")]
+    Unknown,
+}
+
 impl IRRootSignature {
     pub fn create_from_descriptor(
         compiler: &IRCompiler,
         desc: &ffi::IRVersionedRootSignatureDescriptor,
-    ) -> Result<IRRootSignature, Box<dyn std::error::Error>> {
-        unsafe {
-            let mut error = std::ptr::null_mut();
+    ) -> Result<IRRootSignature, RootSignatureError> {
+        let mut error = std::ptr::null_mut();
 
-            let me = compiler
+        let me = unsafe {
+            compiler
                 .funcs
-                .IRRootSignatureCreateFromDescriptor(desc, &mut error);
+                .IRRootSignatureCreateFromDescriptor(desc, &mut error)
+        };
 
-            Ok(Self {
-                me,
-                funcs: compiler.funcs.clone(),
-            })
+        // If the root signature failed to create
+        if !error.is_null() {
+            // IRErrorCode is #[repr(u32)], so this transmute should be fine
+            unsafe {
```

Comment:
You normally start such docs with `// Safety: the explanation`. And this comment expresses 1 line while there are 4 lines in this block, 3 of which are unsafe.

Reply:
Transmuting to an `enum` is anyway never sound: https://doc.rust-lang.org/reference/behavior-considered-undefined.html

You should use the `num_enum` crate or a manual implementation to validate that `code` is one of the valid discriminants of the `enum`.

Reply:
Is it possible to use `num_enum` for an automatically generated bindings.rs file? Otherwise I'm essentially just rewriting the struct again which doesn't make sense in my head

Reply:
Yeah that's always how `num_enum` works, and in that case people typically use a `--newtype-enum` instead of `--rustified-enum` to make it easier to reassign the discriminants.

Otherwise you have to figure out if `bindgen` can generate a `TryFrom<u32> for IRErrorCode`.